File: common/src/main/kotlin/net/impleri/playerskills/items/ItemRestriction.kt
package net.impleri.playerskills.items

import net.impleri.playerskills.restrictions.AbstractRestriction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item

open class ItemRestriction(
  item: Item,
  condition: (Player) -> Boolean,
  replacement: Item? = null,
  includeDimensions: List<ResourceLocation>? = null,
  excludeDimensions: List<ResourceLocation>? = null,
  includeBiomes: List<ResourceLocation>? = null,
  excludeBiomes: List<ResourceLocation>? = null,
  producible: Boolean? = null,
  consumable: Boolean? = null,
  holdable: Boolean? = null,
  identifiable: Boolean? = null,
  harmful: Boolean? = null,
  wearable: Boolean? = null,
  usable: Boolean? = null,
) : AbstractRestriction<Item>(
  item,
  condition,
  includeDimensions ?: ArrayList(),
  excludeDimensions ?: ArrayList(),
  includeBiomes ?: ArrayList(),
  excludeBiomes ?: ArrayList(),
  replacement,
) {
  val producible = producible ?: false
  val consumable = consumable ?: false
  val holdable = holdable ?: false
  val identifiable = identifiable ?: false
  val harmful = harmful ?: false
  val wearable = wearable ?: false
  val usable = usable ?: false
}

File: common/src/main/scala/net/impleri/playerskills/restrictions/item/ItemRestriction.scala
package net.impleri.playerskills.restrictions.item

import net.impleri.playerskills.api.restrictions.Restriction
import net.impleri.playerskills.api.restrictions.RestrictionType
import net.impleri.slab.entity.Player
import net.impleri.slab.item.Item

case class ItemRestriction(
  target: Item,
  condition: Player[_] => Boolean = Restriction.DEFAULT_CONDITION,
  replacement: Option[Item] = None,
  includeDimensions: Seq[String] = List.empty,
  excludeDimensions: Seq[String] = List.empty,
  includeBiomes: Seq[String] = List.empty,
  excludeBiomes: Seq[String] = List.empty,
  holdable: Boolean = Restriction.DEFAULT_RESPONSE,
  wearable: Boolean = Restriction.DEFAULT_RESPONSE,
  harmful: Boolean = Restriction.DEFAULT_RESPONSE,
  usable: Boolean = Restriction.DEFAULT_RESPONSE,
  identifiable: Boolean = Restriction.DEFAULT_RESPONSE,
) extends Restriction[Item, Item.Vanilla] {
  override val restrictionType: RestrictionType = RestrictionType.Item()
}

object ItemRestriction {
  def apply(
    target: Item,
    builder: ItemConditions,
  ): ItemRestriction = {
    new ItemRestriction(
      target,
      builder.condition,
      None,
      builder.includeDimensions,
      builder.excludeDimensions,
      builder.includeBiomes,
      builder.excludeBiomes,
      builder.isHoldable.getOrElse(Restriction.DEFAULT_RESPONSE),
      builder.isWearable.getOrElse(Restriction.DEFAULT_RESPONSE),
      builder.isHarmful.getOrElse(Restriction.DEFAULT_RESPONSE),
      builder.isUsable.getOrElse(Restriction.DEFAULT_RESPONSE),
      builder.isIdentifiable.getOrElse(Restriction.DEFAULT_RESPONSE),
    )
  }
}
